### ISSUE TITLE
Override unsuitable Location header values

### ISSUE THREAD

**Issue by mdr1384** [PostgREST/postgrest#1427]
### Environment

* PostgreSQL version: (if using docker, specify the image)
  11.5
* PostgREST version: (if using docker, specify the image)
  6.0.2
* Operating system:
  FreeBSD 11.3-RELEASE-p5
### Description of issue

For some of our endpoints the Location header does not use the value we want. In some cases there is no Location header returned, and in other cases it uses an unsuitable combination of columns.

In the case where it does not return a Location header, we were able to overcome this by putting the following in our `INSTEAD OF INSERT` trigger function:
```sql
    -- Provide a value for the Location header
    IF NEW.id IS NOT NULL
    THEN
        PERFORM SET_CONFIG(
                'response.headers',
                FORMAT('[{"Location": "/%s?id=eq.%s"}]', TG_TABLE_NAME, NEW.id),
                FALSE
            );
    END IF;
```

The trouble is for the case where PostgREST uses an unsuitable value. In this case the above code does not seem to override the value chosen by PostgREST.  This is especially problematic when the chosen columns contain huge amounts of data, which exceed the 4KB limit for header info.

So far we have not been able to do anything to overcome this, except disable the Location header completely in the web server configs for the affected endpoints.

If PostgREST were to only set the Location header value if it was not already set (assuming PostgREST is setting it _after_ the INSERT happens) -- then it seems there would be a way to override it when the default value was unsuitable.

I looked at the code (src/PostgREST/QueryBuilder/WriteStatement.hs) a bit but Haskell is like egyptian hieroglyphics to me -- I did not see how to do it...

Also maybe consider NOT setting the Location header if the resulting string is more than X bytes long (2000? 3000?).

**Comment by steve-chavez** [comment]
@mdr1384 Overriding the `Location` header should be possible. I agree. 

Your trigger snippet is a bit surprising. Does that add the response header when POSTing? I remember we had to clear #1088 for that to work(adding headers only works on functions calls for now).

> Also maybe consider NOT setting the Location header if the resulting string is more than X bytes long (2000? 3000?)

This is something that also happened to me when inserting a `bytea`, IIRC curl erred when this happened. We should handle this case.

**Comment by mdr1384** [comment]
> Does that add the response header when POSTing?
Seems to. We have some endpoints which are views containing complex joins and sub-queries that have "instead of" triggers for insert, update and delete - which for whatever reason did not generate a Location header upon POSTing. I added the code in the snippet to the insert trigger and the Location header was restored.  However, later I tried to figure out why it was missing in the first place and I discovered that removing the sub-queries from the SELECT clause and adding them as LATERAL JOINs also restored the Location header, without the code snippet.

**Comment by mdr1384** [comment]
Hi - I was excited to try your change to allow setting `response.headers.location` - so I grabbed [master/de218e9](https://github.com/PostgREST/postgrest/commit/de218e900b4d30b98d2508a71045dadff25118a3) and tried it... but the case where I was getting an unsuitable location (e.g. it included ALL columns rather than just the primary key) -- it is now returning a 500 status code (although it DOES create the entry.)

```
Request Headers
---------------
Prefer: return=representation
Content-Type: application/json
Accept: application/json
Authorization: Bearer eyJ...
User-Agent: PostmanRuntime/7.22.0
Cache-Control: no-cache
Postman-Token: 66c76815-b6eb-4712-807d-de359d86a2e2
Host: florida.redcom.com
Accept-Encoding: gzip, deflate, br
Content-Length: 28
Connection: keep-alive

Request Body
------------
{"pattern": "4569123"}

Response Code
-------------
500 Internal Server Error

Response Headers
----------------
Date: Mon, 02 Mar 2020 19:43:49 GMT
Server: Apache
Connection: close
Content-Type: text/html; charset=iso-8859-1

Response Body
-------------
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>500 Internal Server Error</title>
</head><body>
<h1>Internal Server Error</h1>
<p>The server encountered an internal error or
misconfiguration and was unable to complete
your request.</p>
<p>Please contact the server administrator at
 you@example.com to inform them of the time this error occurred,
 and the actions you performed just before this error.</p>
<p>More information about this error may be available
in the server error log.</p>
<p>Additionally, a 201 Created
error was encountered while trying to use an ErrorDocument to handle the request.</p>
</body></html>
```

Also the apache server log has an error in it: `AH02428: Empty response header name, aborting request`

I tried logging the response headers using mod_security `SecAuditLogParts ABIJDFHZ`, but it shows the same thing that Postman sees - it can't show the intended response headers, only the actually generated ones, but I suspect it is the Location header that is broken, because the DELETE and PATCH methods on this endpoint still work - it is only the POST method that broke.

I also tried overriding the Location header using my `SET_CONFIG()` code as shown in the O.P. but it does not change anything.

**Comment by steve-chavez** [comment]
@mdr1384 That's strange, because I added a test that has an use case similar to yours(overriding Location on trigger):

https://github.com/PostgREST/postgrest/blob/de218e900b4d30b98d2508a71045dadff25118a3/test/fixtures/schema.sql#L1673-L1690

And it has been running succesfully:

https://github.com/PostgREST/postgrest/blob/de218e900b4d30b98d2508a71045dadff25118a3/test/Feature/PgVersion96Spec.hs#L128-L133

Seems that perhaps your header value is malformed(empty maybe?) and Apache is erring because of that. I would suggest isolating the problem from Apache first and checking the generated Location values.

**Comment by mdr1384** [comment]
Well, here is my exact code:

```
    PERFORM SET_CONFIG(
            'response.headers',
            FORMAT('[{"Location": "/%s?port_id=eq.%s"}]', TG_TABLE_NAME, NEW.port_id),
            TRUE
        );
```

It looks right to me.  Is the `&overridden=true` necessary? Because we don't want that coming out to the user. Also I note you are using `FALSE` for the "is_local" parameter. Not sure why that would matter... I tried it both ways and no difference. [The docs](https://www.postgresql.org/docs/11/functions-admin.html) say "If is_local is true, the new value will only apply to the current transaction. If you want the new value to apply for the current session, use false instead."

BTW - I spent a great deal of time trying to intercept and print out the headers between postgrest and apache (using modsecurity) and cannot find any way to do it. You can't observe the outgoing headers until after apache is done whacking on them.  Do you know of some way to see the "intended" headers?

**Comment by steve-chavez** [comment]
> Is the &overridden=true necessary?

No. That is just an example for the test suite to check.

> Also I note you are using FALSE for the "is_local" parameter.

Yeah, I should have used true there. But there won't be a difference for the test.

> Do you know of some way to see the "intended" headers?

You could ssh into the box and make the request locally to postgrest(`curl -v ... localhost:3000`). Then you'll see if it's indeed responding with the correct headers.

**Comment by mdr1384** [comment]
> You could ssh into the box and make the request locally

Oh, yeah, good idea... (smacking head).  So, it gives me:

```
*   Trying 127.0.0.1:3000...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 3000 (#0)
> POST /loc_test HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.66.0
> Accept: application/json
> Authorization: Bearer eyJ...
> Connection: keep-alive
> Content-Type: application/json
> Content-Length: 49
>
* upload completely sent off: 49 out of 49 bytes
* Mark bundle as not supporting multiuse
< HTTP/1.1 201 Created
HTTP/1.1 201 Created
< Transfer-Encoding: chunked
Transfer-Encoding: chunked
< Date: Thu, 05 Mar 2020 17:54:08 GMT
Date: Thu, 05 Mar 2020 17:54:08 GMT
< Server: postgrest/6.0.2 (UNKNOWN)
Server: postgrest/6.0.2 (UNKNOWN)
< :
:
< Content-Range: */*
Content-Range: */*
< :
:
< Location: /loc_test?port_id=eq.205
Location: /loc_test?port_id=eq.205

<
* Connection #0 to host localhost left intact
```

Looks like it is trying to set the Location header, but for some reason apache doesn't like it?

**Comment by steve-chavez** [comment]
Apache could reject headers values with whitespace: https://blog.tigertech.net/posts/apache-cve-2016-8743/. Also see https://serverfault.com/questions/853103/500-internal-server-ah02429-response-header-name.

Maybe try different headers key-value combinations and see what could be wrong.

**Comment by mdr1384** [comment]
It looks to me like PostgREST is generating two headers with no name and no value, thus ending up with ":" only. It is nothing our schema code is doing, because I can run our schema with the released 6.0.2 and get a valid response with an unsuitable Location header:

```
*   Trying 127.0.0.1:3000...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 3000 (#0)
> POST /loc_test HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.66.0
> Accept: application/json
> Authorization: Bearer eyJ...
> Content-Type: application/json
> Content-Length: 49
>
* upload completely sent off: 49 out of 49 bytes
* Mark bundle as not supporting multiuse
< HTTP/1.1 201 Created
< Transfer-Encoding: chunked
< Date: Thu, 05 Mar 2020 18:57:54 GMT
< Server: postgrest/6.0.2 (UNKNOWN)
< Location: /loc_test?port_id=eq.206&name=eq.Loc%20header%20test&...(2k+ chars)
< Content-Range: */*
<
* Connection #0 to host localhost left intact
```
Then if I swap in the latest PostgREST version master/de218e9, with no other changes, I get the bad response:

```
*   Trying 127.0.0.1:3000...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 3000 (#0)
> POST /lines_sip_quick HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.66.0
> Accept: application/json
> Authorization: Bearer eyJ...
> Content-Type: application/json
> Content-Length: 49
>
* upload completely sent off: 49 out of 49 bytes
* Mark bundle as not supporting multiuse
< HTTP/1.1 201 Created
< Transfer-Encoding: chunked
< Date: Thu, 05 Mar 2020 19:06:32 GMT
< Server: postgrest/6.0.2 (UNKNOWN)
< :
< Content-Range: */*
< :
< Location: /lines_sip_quick?port_id=eq.207
<
* Connection #0 to host localhost left intact
```

> Maybe try different headers key-value combinations and see what could be wrong.

I tried a few things, and it all comes back to PostgREST trying to set two headers without a name or a value, all I can tell is that it is not the "Location" header it is trying to set, because if I don't override it, then the old unsuitable location header comes out along with the two empty headers. I also tried showing what headers were set at the time of my code being executed (which is at the very end of the trigger function so there is no other code coming after):

```
    PERFORM SET_CONFIG(
            'response.headers',
            FORMAT('[{"BLocation": "/%s?port_id=eq.%s"}]', TG_TABLE_NAME, NEW.port_id),
            TRUE
        );
```
```
< HTTP/1.1 201 Created
< Transfer-Encoding: chunked
< Date: Thu, 05 Mar 2020 19:10:06 GMT
< Server: postgrest/6.0.2 (UNKNOWN)
< Location: /lines_sip_quick?port_id=eq.211&name=eq.Loc%20header%20test& . . .
< :
< Content-Range: */*
< :
< BLocation: /lines_sip_quick?port_id=eq.211
<
```
```
    PERFORM SET_CONFIG(
            'response.headers',
            FORMAT('[{"Recursive": "%s"}]', CURRENT_SETTING('response.headers', TRUE)::TEXT),
            TRUE
        );
```
```
< HTTP/1.1 201 Created
< Transfer-Encoding: chunked
< Date: Thu, 05 Mar 2020 19:30:03 GMT
< Server: postgrest/6.0.2 (UNKNOWN)
< Location: /lines_sip_quick?port_id=eq.211&name=eq.Loc%20header%20test& . . .
< :
< Content-Range: */*
< :
< Recursive:
<
```

Something strange also - it seems you can only do the SET_CONFIG once. E.g.:

```
    PERFORM SET_CONFIG(
            'response.headers',
            FORMAT('[{"BLocation": "/%s?port_id=eq.%s"}]', TG_TABLE_NAME, NEW.port_id),
            TRUE
        );
    PERFORM SET_CONFIG(
            'response.headers',
            FORMAT('[{"Recursive": "%s"}]', CURRENT_SETTING('response.headers', TRUE)::TEXT),
            TRUE
        );
```

Blows up with:

```
< HTTP/1.1 500 Internal Server Error
< Transfer-Encoding: chunked
< Date: Thu, 05 Mar 2020 19:36:21 GMT
< Server: postgrest/6.0.2 (UNKNOWN)
< Content-Type: application/json; charset=utf-8
<
{"message":"response.headers guc must be a JSON array composed of objects with a single key and a string value"}
```

**Comment by steve-chavez** [comment]
> because I can run our schema with the released 6.0.2 and get a valid response with an unsuitable Location header

> Location: /loc_test?port_id=eq.206&name=eq.Loc%20header%20test&...(2k+ chars)

The large Location header is indeed a problem we should fix, I forgot to correct this so I'll reopen.

> Blows up with:

The last error message you got means that the JSON was malformed. I can set two different headers with two PERFORMs and it works normally.

It can be error prone to type the JSON, using `json_build_array` like this would be better:
```sql
  headers := json_build_array(
    json_build_object('Content-type', 'application/octet-stream'),
    json_build_object('Content-Transfer-Encoding', 'binary'),
    json_build_object('Content-Disposition', 'attachment; filename=report.zip'),
    json_build_object('Content-Length', octet_length(result)::text),
    json_build_object('Access-Control-Expose-Headers', 'Content-Disposition'),
    json_build_object('Cache-control', 'no-cache')
  );
  perform set_config('response.headers', headers::text, true);
```

**Comment by mdr1384** [comment]
We're getting off track.  The main issue I am having is that PostgREST is adding invalid headers in the response, which are not a result of us doing SET_CONFIG().  There are 2 of them and both are showing up as empty name and empty value (or possibly spaces only.)  It is only doing it with "master" version, not with released 6.0.2.  And after a while of incrementally simplifying my schema, I found that even the very simplest schema exhibits the problem:

```pgsql
CREATE TABLE IF NOT EXISTS api.loc_test (
    id SERIAL PRIMARY KEY,
    c TEXT
);
GRANT ALL PRIVILEGES ON api.loc_test TO rest_user;
GRANT ALL PRIVILEGES ON SEQUENCE api.loc_test_id_seq TO rest_user;
```

Version 6.0.2 (ok):

```
curl --trace - -H 'Authorization: Bearer eyJ...' -H 'Content-Type: application/json' \
    --data '{"c":"c1"}' 'localhost:3000/loc_test'
== Info:   Trying 127.0.0.1:3000...
== Info: TCP_NODELAY set
== Info: Connected to localhost (127.0.0.1) port 3000 (#0)
=> Send header, 654 bytes (0x28e)
0000: 50 4f 53 54 20 2f 6c 6f 63 5f 74 65 73 74 20 48 POST /loc_test H
0010: 54 54 50 2f 31 2e 31 0d 0a 48 6f 73 74 3a 20 6c TTP/1.1..Host: l
0020: 6f 63 61 6c 68 6f 73 74 3a 33 30 30 30 0d 0a 55 ocalhost:3000..U
0030: 73 65 72 2d 41 67 65 6e 74 3a 20 63 75 72 6c 2f ser-Agent: curl/
0040: 37 2e 36 36 2e 30 0d 0a 41 63 63 65 70 74 3a 20 7.66.0..Accept:
0050: 2a 2f 2a 0d 0a 41 75 74 68 6f 72 69 7a 61 74 69 */*..Authorizati
0060: 6f 6e 3a 20 42 65 61 72 65 72 20 65 79 4a 68 62 on: Bearer eyJhb
0070: 47 63 69 4f 69 4a 49 55 7a 49 31 4e 69 49 73 49 GciOiJIUzI1NiIsI
0080: 6e 52 35 63 43 49 36 49 6b 70 58 56 43 4a 39 2e nR5cCI6IkpXVCJ9.
0090: 65 79 4a 71 64 47 6b 69 49 44 6f 67 49 6e 6b 34 eyJqdGkiIDogInk4
00a0: 57 55 38 30 54 30 64 51 57 6d 74 32 61 6b 68 56 WU80T0dQWmt2akhV
00b0: 5a 43 39 79 51 7a 59 31 62 58 41 7a 52 6e 5a 55 ZC9yQzY1bXAzRnZU
00c0: 65 45 70 70 63 30 39 6b 62 46 6c 4d 64 57 56 53 eEppc09kbFlMdWVS
00d0: 64 54 52 52 63 31 55 39 49 69 77 67 49 6d 6c 7a dTRRc1U9IiwgImlz
00e0: 63 79 49 67 4f 69 41 69 5a 6d 78 76 63 6d 6c 6b cyIgOiAiZmxvcmlk
00f0: 59 53 35 79 5a 57 52 6a 62 32 30 75 59 32 39 74 YS5yZWRjb20uY29t
0100: 49 69 77 67 49 6d 6c 68 64 43 49 67 4f 69 41 78 IiwgImlhdCIgOiAx
0110: 4e 54 67 7a 4e 54 45 32 4e 7a 41 32 4c 43 41 69 NTgzNTE2NzA2LCAi
0120: 62 6d 4a 6d 49 69 41 36 49 44 45 31 4f 44 4d 31 bmJmIiA6IDE1ODM1
0130: 4d 54 59 32 4e 7a 59 73 49 43 4a 6c 65 48 41 69 MTY2NzYsICJleHAi
0140: 49 44 6f 67 4d 54 55 34 4d 7a 55 78 4e 7a 41 79 IDogMTU4MzUxNzAy
0150: 4e 69 77 67 49 6e 4a 6c 5a 47 4e 76 62 56 39 7a NiwgInJlZGNvbV9z
0160: 61 57 64 74 59 56 39 79 5a 58 4e 30 58 33 4a 34 aWdtYV9yZXN0X3J4
0170: 63 43 49 67 4f 69 41 78 4e 54 67 7a 4e 54 49 7a cCIgOiAxNTgzNTIz
0180: 4f 54 49 32 4c 43 41 69 63 6d 56 6b 59 32 39 74 OTI2LCAicmVkY29t
0190: 58 33 4e 70 5a 32 31 68 58 33 4a 6c 63 33 52 66 X3NpZ21hX3Jlc3Rf
01a0: 64 58 4e 6c 63 6c 39 70 5a 43 49 67 4f 69 41 78 dXNlcl9pZCIgOiAx
01b0: 4c 43 41 69 63 6d 56 6b 59 32 39 74 58 33 4e 70 LCAicmVkY29tX3Np
01c0: 5a 32 31 68 58 33 4a 6c 63 33 52 66 62 57 39 6b Z21hX3Jlc3RfbW9k
01d0: 5a 56 39 70 5a 43 49 67 4f 69 41 77 4c 43 41 69 ZV9pZCIgOiAwLCAi
01e0: 63 6d 56 6b 59 32 39 74 58 33 4e 70 5a 32 31 68 cmVkY29tX3NpZ21h
01f0: 58 33 4a 6c 63 33 52 66 64 6e 4e 66 61 57 51 69 X3Jlc3RfdnNfaWQi
0200: 49 44 6f 67 4d 43 77 67 49 6e 4a 76 62 47 55 69 IDogMCwgInJvbGUi
0210: 49 44 6f 67 49 6e 4a 6c 63 33 52 66 64 58 4e 6c IDogInJlc3RfdXNl
0220: 63 6c 39 74 59 58 67 69 66 51 2e 32 54 55 47 74 cl9tYXgifQ.2TUGt
0230: 44 62 68 64 49 33 66 5a 31 71 30 4c 66 6f 69 51 DbhdI3fZ1q0LfoiQ
0240: 6c 6d 77 6d 4c 70 71 49 4f 46 36 38 74 34 6c 6e lmwmLpqIOF68t4ln
0250: 43 61 7a 6d 55 49 0d 0a 43 6f 6e 74 65 6e 74 2d CazmUI..Content-
0260: 54 79 70 65 3a 20 61 70 70 6c 69 63 61 74 69 6f Type: applicatio
0270: 6e 2f 6a 73 6f 6e 0d 0a 43 6f 6e 74 65 6e 74 2d n/json..Content-
0280: 4c 65 6e 67 74 68 3a 20 31 30 0d 0a 0d 0a       Length: 10....
=> Send data, 10 bytes (0xa)
0000: 7b 22 63 22 3a 22 63 31 22 7d                   {"c":"c1"}
== Info: upload completely sent off: 10 out of 10 bytes
== Info: Mark bundle as not supporting multiuse
<= Recv header, 22 bytes (0x16)
0000: 48 54 54 50 2f 31 2e 31 20 32 30 31 20 43 72 65 HTTP/1.1 201 Cre
0010: 61 74 65 64 0d 0a                               ated..
<= Recv header, 28 bytes (0x1c)
0000: 54 72 61 6e 73 66 65 72 2d 45 6e 63 6f 64 69 6e Transfer-Encodin
0010: 67 3a 20 63 68 75 6e 6b 65 64 0d 0a             g: chunked..
<= Recv header, 37 bytes (0x25)
0000: 44 61 74 65 3a 20 46 72 69 2c 20 30 36 20 4d 61 Date: Fri, 06 Ma
0010: 72 20 32 30 32 30 20 31 37 3a 34 35 3a 31 36 20 r 2020 17:45:16
0020: 47 4d 54 0d 0a                                  GMT..
<= Recv header, 35 bytes (0x23)
0000: 53 65 72 76 65 72 3a 20 70 6f 73 74 67 72 65 73 Server: postgres
0010: 74 2f 36 2e 30 2e 32 20 28 55 4e 4b 4e 4f 57 4e t/6.0.2 (UNKNOWN
0020: 29 0d 0a                                        )..
<= Recv header, 37 bytes (0x25)
0000: 4c 6f 63 61 74 69 6f 6e 3a 20 2f 6c 6f 63 5f 74 Location: /loc_t
0010: 65 73 74 3f 69 64 3d 65 71 2e 31 26 63 3d 65 71 est?id=eq.1&c=eq
0020: 2e 63 31 0d 0a                                  .c1..
<= Recv header, 20 bytes (0x14)
0000: 43 6f 6e 74 65 6e 74 2d 52 61 6e 67 65 3a 20 2a Content-Range: *
0010: 2f 2a 0d 0a                                     /*..
<= Recv header, 2 bytes (0x2)
0000: 0d 0a                                           ..
<= Recv data, 5 bytes (0x5)
0000: 30 0d 0a 0d 0a                                  0....
== Info: Connection #0 to host localhost left intact
```

Version master-de218e9 (ng):
```
curl --trace - -H 'Authorization: Bearer eyJ...' -H 'Content-Type: application/json' \
     --data '{"c":"c1"}' 'localhost:3000/loc_test'
== Info:   Trying 127.0.0.1:3000...
== Info: TCP_NODELAY set
== Info: Connected to localhost (127.0.0.1) port 3000 (#0)
=> Send header, 654 bytes (0x28e)
0000: 50 4f 53 54 20 2f 6c 6f 63 5f 74 65 73 74 20 48 POST /loc_test H
0010: 54 54 50 2f 31 2e 31 0d 0a 48 6f 73 74 3a 20 6c TTP/1.1..Host: l
0020: 6f 63 61 6c 68 6f 73 74 3a 33 30 30 30 0d 0a 55 ocalhost:3000..U
0030: 73 65 72 2d 41 67 65 6e 74 3a 20 63 75 72 6c 2f ser-Agent: curl/
0040: 37 2e 36 36 2e 30 0d 0a 41 63 63 65 70 74 3a 20 7.66.0..Accept:
0050: 2a 2f 2a 0d 0a 41 75 74 68 6f 72 69 7a 61 74 69 */*..Authorizati
0060: 6f 6e 3a 20 42 65 61 72 65 72 20 65 79 4a 68 62 on: Bearer eyJhb
0070: 47 63 69 4f 69 4a 49 55 7a 49 31 4e 69 49 73 49 GciOiJIUzI1NiIsI
0080: 6e 52 35 63 43 49 36 49 6b 70 58 56 43 4a 39 2e nR5cCI6IkpXVCJ9.
0090: 65 79 4a 71 64 47 6b 69 49 44 6f 67 49 6c 46 45 eyJqdGkiIDogIlFE
00a0: 51 6d 46 74 54 57 68 69 63 58 4a 42 4e 57 38 30 QmFtTWhicXJBNW80
00b0: 55 6b 74 4b 56 6e 52 6c 65 55 5a 42 64 32 4e 56 UktKVnRleUZBd2NV
00c0: 4f 46 4e 6d 4f 56 6c 4f 53 30 70 6b 52 32 49 7a OFNmOVlOS0pkR2Iz
00d0: 64 57 4d 72 4f 45 55 39 49 69 77 67 49 6d 6c 7a dWMrOEU9IiwgImlz
00e0: 63 79 49 67 4f 69 41 69 5a 6d 78 76 63 6d 6c 6b cyIgOiAiZmxvcmlk
00f0: 59 53 35 79 5a 57 52 6a 62 32 30 75 59 32 39 74 YS5yZWRjb20uY29t
0100: 49 69 77 67 49 6d 6c 68 64 43 49 67 4f 69 41 78 IiwgImlhdCIgOiAx
0110: 4e 54 67 7a 4e 54 45 32 4f 54 49 77 4c 43 41 69 NTgzNTE2OTIwLCAi
0120: 62 6d 4a 6d 49 69 41 36 49 44 45 31 4f 44 4d 31 bmJmIiA6IDE1ODM1
0130: 4d 54 59 34 4f 54 41 73 49 43 4a 6c 65 48 41 69 MTY4OTAsICJleHAi
0140: 49 44 6f 67 4d 54 55 34 4d 7a 55 78 4e 7a 49 30 IDogMTU4MzUxNzI0
0150: 4d 43 77 67 49 6e 4a 6c 5a 47 4e 76 62 56 39 7a MCwgInJlZGNvbV9z
0160: 61 57 64 74 59 56 39 79 5a 58 4e 30 58 33 4a 34 aWdtYV9yZXN0X3J4
0170: 63 43 49 67 4f 69 41 78 4e 54 67 7a 4e 54 49 30 cCIgOiAxNTgzNTI0
0180: 4d 54 51 77 4c 43 41 69 63 6d 56 6b 59 32 39 74 MTQwLCAicmVkY29t
0190: 58 33 4e 70 5a 32 31 68 58 33 4a 6c 63 33 52 66 X3NpZ21hX3Jlc3Rf
01a0: 64 58 4e 6c 63 6c 39 70 5a 43 49 67 4f 69 41 78 dXNlcl9pZCIgOiAx
01b0: 4c 43 41 69 63 6d 56 6b 59 32 39 74 58 33 4e 70 LCAicmVkY29tX3Np
01c0: 5a 32 31 68 58 33 4a 6c 63 33 52 66 62 57 39 6b Z21hX3Jlc3RfbW9k
01d0: 5a 56 39 70 5a 43 49 67 4f 69 41 77 4c 43 41 69 ZV9pZCIgOiAwLCAi
01e0: 63 6d 56 6b 59 32 39 74 58 33 4e 70 5a 32 31 68 cmVkY29tX3NpZ21h
01f0: 58 33 4a 6c 63 33 52 66 64 6e 4e 66 61 57 51 69 X3Jlc3RfdnNfaWQi
0200: 49 44 6f 67 4d 43 77 67 49 6e 4a 76 62 47 55 69 IDogMCwgInJvbGUi
0210: 49 44 6f 67 49 6e 4a 6c 63 33 52 66 64 58 4e 6c IDogInJlc3RfdXNl
0220: 63 6c 39 74 59 58 67 69 66 51 2e 66 56 6c 70 67 cl9tYXgifQ.fVlpg
0230: 4d 34 61 72 46 38 47 72 6b 4b 5a 5a 7a 62 71 37 M4arF8GrkKZZzbq7
0240: 54 44 36 32 70 51 31 59 41 52 72 57 42 49 6e 56 TD62pQ1YARrWBInV
0250: 49 36 39 6e 63 34 0d 0a 43 6f 6e 74 65 6e 74 2d I69nc4..Content-
0260: 54 79 70 65 3a 20 61 70 70 6c 69 63 61 74 69 6f Type: applicatio
0270: 6e 2f 6a 73 6f 6e 0d 0a 43 6f 6e 74 65 6e 74 2d n/json..Content-
0280: 4c 65 6e 67 74 68 3a 20 31 30 0d 0a 0d 0a       Length: 10....
=> Send data, 10 bytes (0xa)
0000: 7b 22 63 22 3a 22 63 31 22 7d                   {"c":"c1"}
== Info: upload completely sent off: 10 out of 10 bytes
== Info: Mark bundle as not supporting multiuse
<= Recv header, 22 bytes (0x16)
0000: 48 54 54 50 2f 31 2e 31 20 32 30 31 20 43 72 65 HTTP/1.1 201 Cre
0010: 61 74 65 64 0d 0a                               ated..
<= Recv header, 28 bytes (0x1c)
0000: 54 72 61 6e 73 66 65 72 2d 45 6e 63 6f 64 69 6e Transfer-Encodin
0010: 67 3a 20 63 68 75 6e 6b 65 64 0d 0a             g: chunked..
<= Recv header, 37 bytes (0x25)
0000: 44 61 74 65 3a 20 46 72 69 2c 20 30 36 20 4d 61 Date: Fri, 06 Ma
0010: 72 20 32 30 32 30 20 31 37 3a 34 38 3a 34 39 20 r 2020 17:48:49
0020: 47 4d 54 0d 0a                                  GMT..
<= Recv header, 35 bytes (0x23)
0000: 53 65 72 76 65 72 3a 20 70 6f 73 74 67 72 65 73 Server: postgres
0010: 74 2f 36 2e 30 2e 32 20 28 55 4e 4b 4e 4f 57 4e t/6.0.2 (UNKNOWN
0020: 29 0d 0a                                        )..
<= Recv header, 37 bytes (0x25)
0000: 4c 6f 63 61 74 69 6f 6e 3a 20 2f 6c 6f 63 5f 74 Location: /loc_t
0010: 65 73 74 3f 69 64 3d 65 71 2e 31 26 63 3d 65 71 est?id=eq.1&c=eq
0020: 2e 63 31 0d 0a                                  .c1..
<= Recv header, 4 bytes (0x4)
0000: 3a 20 0d 0a                                     : ..
<= Recv header, 20 bytes (0x14)
0000: 43 6f 6e 74 65 6e 74 2d 52 61 6e 67 65 3a 20 2a Content-Range: *
0010: 2f 2a 0d 0a                                     /*..
<= Recv header, 4 bytes (0x4)
0000: 3a 20 0d 0a                                     : ..
<= Recv header, 2 bytes (0x2)
0000: 0d 0a                                           ..
<= Recv data, 5 bytes (0x5)
0000: 30 0d 0a 0d 0a                                  0....
== Info: Connection #0 to host localhost left intact
```

I cannot possibly imagine what external thing could be causing that.

**Comment by steve-chavez** [comment]
> There are 2 of them and both are showing up as empty name and empty value (or possibly spaces only)

What are the 2 headers? Please make an easily reproducible test case.

**Comment by steve-chavez** [comment]
@mdr1384 You're right, thanks for catching this. I can reproduce the issue now.

```bash
curl -i -H 'Content-Type: application/json' --data '{"c":"c2"}' \
 'localhost:3000/loc_test'

HTTP/1.1 201 Created
Transfer-Encoding: chunked
Date: Fri, 06 Mar 2020 18:47:06 GMT
Server: postgrest/6.0.2 (0db2d0b)
Location: /loc_test?id=eq.17
:
Content-Range: */*
:
```
Two empty headers(`:`) are being added.

**Comment by mdr1384** [comment]
Hey, how come in your test the Location header is correct? In my case the Location header was wrong.
```
curl -i -H 'Authorization: Bearer eyJ...' -H 'Content-Type: application/json' \
    --data '{"c":"c1"}' 'localhost:3000/loc_test'
HTTP/1.1 201 Created
Transfer-Encoding: chunked
Date: Fri, 06 Mar 2020 19:51:32 GMT
Server: postgrest/6.0.2 (UNKNOWN)
Location: /loc_test?id=eq.1&c=eq.c1
:
Content-Range: */*
:
```

**Comment by steve-chavez** [comment]
@mdr1384 The `Location: /loc_test?id=eq.1&c=eq.c1` happens when the `id` is not defined as primary key(also if the pk constraint has just been added and the schema cache is not updated).

If there's a primary key, we only add the pk column to the querystring `/loc_test?id=eq.1` for the `Location` header, if not we add all of the columns.

I've fixed the empty headers on https://github.com/PostgREST/postgrest/pull/1458. Maybe you can try it to double check the empty headers are gone.

**Comment by mdr1384** [comment]
> Maybe you can try it to double check the empty headers are gone.

I want to, but having trouble at the moment finding a machine where I can build from source...

**Comment by mdr1384** [comment]
@steve-chavez - I was able to verify your change for #1458 is working. I could override the Location header, without any empty headers being added.

steve-chavez:empty-headers:
```
HTTP/1.1 201 Created
Transfer-Encoding: chunked
Date: Wed, 11 Mar 2020 14:26:57 GMT
Server: postgrest/6.0.2 (UNKNOWN)
Content-Range: */*
Location: /lines_sip_quick?port_id=eq.226
```

postgrest:master:
```
HTTP/1.1 201 Created
Transfer-Encoding: chunked
Date: Wed, 11 Mar 2020 14:28:42 GMT
Server: postgrest/6.0.2 (UNKNOWN)
:
Content-Range: */*
:
Location: /lines_sip_quick?port_id=eq.227
```

postgrest:6.0.2:
```
HTTP/1.1 201 Created
Transfer-Encoding: chunked
Date: Wed, 11 Mar 2020 14:29:27 GMT
Server: postgrest/6.0.2 (UNKNOWN)
Location: /lines_sip_quick?port_id=eq.228&name=eq.Postman%20Test%20Quick%20Line&...(45 columns)...
Content-Range: */*
```

**Comment by steve-chavez** [comment]
@mdr1384 Good! I'll merge the PR then.

Also, does your `lines_sip_quick` table have a primary key?

I was thinking that we should only add the Location header when there's a PK or a UNIQUE constraint that identifies the row.

**Comment by mdr1384** [comment]
> does your `lines_sip_quick` table have a primary key?

It is not a table. All our endpoints are VIEWS which refer to various tables in a different schema. The main underlying table for the `lines_sip_quick` view is `port` which has a SERIAL PK called `port_id` that is selected directly in the view, but there are many other tables involved and although they all do have primary keys, none of those other PKs are selected **_directly_** into the view. Some of the other tables do have their PKs in the view, but as part of an aggregate column. E.g. the `tag` table (with PK `tag_id`) is selected in the view as column called `tags` which is a JSON aggregate including the tag ids and names of all the tags assigned to the line: `[{"id":23,"name":"WORK"},{"id":46,"name":"STAFF"}]`.  The views are complex enough that we have to use INSTEAD OF triggers on all of them for every method.  With our complex views we are having a lot of trouble with UPSERT and with CSV.

BTW - Thank you very much for hanging in there with this Location header thing, it will be a relief to be able to ensure the Location we want in our INSTEAD OF triggers for these complex cases.

**Comment by steve-chavez** [comment]
> BTW - Thank you very much for hanging in there with this Location header thing, it will be a relief to be able to ensure the Location we want in our INSTEAD OF triggers for these complex cases.

@mdr1384 Cool! Thanks for the good bug report. Btw, you can always [support PostgREST on Patreon](https://www.patreon.com/postgrest) to make sure it keeps improving and bugs keep getting fixed promptly :)

For now I've merged #1458, so I'll close this issue since now it has a workaround.

For avoiding the bad default of responding with the bad Location header I've opened https://github.com/PostgREST/postgrest/issues/1461.
